Redeclaring global variables should trigger "no-redeclare", not "no-undef".

Minimal repro case:

```js
/* eslint no-undef: 2, no-redeclare: 2 */
var toString = 1;
```

Expected:

```
/home/isiah/dev-projects/js/ttyped/tmp.js
  2:5  error  "toString" is already defined  no-redeclare

✖ 1 problem (1 error, 0 warnings)
```

Found:

```
path/to/file.js
  2:5  error  "toString" is read only  no-undef

✖ 1 problem (1 error, 0 warnings)
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28486767-redeclaring-global-variables-should-trigger-no-redeclare-not-no-undef?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/lib/rules/no-undef.js
+++ b/lib/rules/no-undef.js
@@ -62,8 +62,7 @@ function hasTypeOfOperator(node) {
 
 module.exports = function(context) {
 
-    var NOT_DEFINED_MESSAGE = "\"{{name}}\" is not defined.",
-        READ_ONLY_MESSAGE = "\"{{name}}\" is read only.";
+    var NOT_DEFINED_MESSAGE = "\"{{name}}\" is not defined.";
 
     var options = context.options[0];
     var considerTypeOf = options && options.typeof === true || false;
@@ -84,8 +83,6 @@ module.exports = function(context) {
 
                 if (!variable) {
                     context.report(ref.identifier, NOT_DEFINED_MESSAGE, { name: name });
-                } else if (ref.isWrite() && variable.writeable === false) {
-                    context.report(ref.identifier, READ_ONLY_MESSAGE, { name: name });
                 }
             });
 

--- a/tests/lib/cli.js
+++ b/tests/lib/cli.js
@@ -501,7 +501,7 @@ describe("cli", function() {
     describe("when executing with global flag", function() {
         it("should default defined variables to read-only", function() {
             var filePath = getFixturePath("undef.js");
-            var exit = cli.execute("--global baz,bat --no-ignore --rule no-undef:2 " + filePath);
+            var exit = cli.execute("--global baz,bat --no-ignore --rule no-native-reassign:2 " + filePath);
 
             assert.isTrue(log.info.calledOnce);
             assert.equal(exit, 1);

--- a/tests/lib/rules/no-native-reassign.js
+++ b/tests/lib/rules/no-native-reassign.js
@@ -45,6 +45,13 @@ ruleTester.run("no-native-reassign", rule, {
             code: "require = 0;",
             env: {node: true},
             errors: [{ message: "require is a read-only native object.", type: "Identifier"}]
-        }
+        },
+
+        // Notifications of readonly are moved from no-undef: https://github.com/eslint/eslint/issues/4504
+        { code: "/*global b:false*/ function f() { b = 1; }", errors: [{ message: "b is a read-only native object.", type: "Identifier"}] },
+        { code: "function f() { b = 1; }", global: { b: false }, errors: [{ message: "b is a read-only native object.", type: "Identifier"}] },
+        { code: "/*global b:false*/ function f() { b++; }", errors: [{ message: "b is a read-only native object.", type: "Identifier"}] },
+        { code: "/*global b*/ b = 1;", errors: [{ message: "b is a read-only native object.", type: "Identifier"}] },
+        { code: "Array = 1;", errors: [{ message: "Array is a read-only native object.", type: "Identifier"}] }
     ]
 });

--- a/tests/lib/rules/no-redeclare.js
+++ b/tests/lib/rules/no-redeclare.js
@@ -93,6 +93,15 @@ ruleTester.run("no-redeclare", rule, {
             errors: [
                 { message: "\"a\" is already defined", type: "Identifier"}
             ]
+        },
+
+        // Notifications of readonly are moved from no-undef: https://github.com/eslint/eslint/issues/4504
+        {
+            code: "/*global b:false*/ var b = 1;",
+            options: [{builtinGlobals: true}],
+            errors: [
+                { message: "\"b\" is already defined", type: "Identifier"}
+            ]
         }
     ]
 });

--- a/tests/lib/rules/no-undef.js
+++ b/tests/lib/rules/no-undef.js
@@ -54,6 +54,14 @@ ruleTester.run("no-undef", rule, {
         { code: "var a; ({b: a}) = {};", ecmaFeatures: {destructuring: true} },
         { code: "var obj; [obj.a, obj.b] = [0, 1];", ecmaFeatures: {destructuring: true} },
 
+        // Notifications of readonly are removed: https://github.com/eslint/eslint/issues/4504
+        { code: "/*global b:false*/ function f() { b = 1; }" },
+        { code: "function f() { b = 1; }", global: { b: false } },
+        { code: "/*global b:false*/ function f() { b++; }" },
+        { code: "/*global b*/ b = 1;" },
+        { code: "/*global b:false*/ var b = 1;" },
+        { code: "Array = 1;" },
+
         // Experimental,
         {
             code: "var {bacon, ...others} = stuff; foo(others)",
@@ -66,14 +74,8 @@ ruleTester.run("no-undef", rule, {
         { code: "if (typeof anUndefinedVar === 'string') {}", options: [{typeof: true}], errors: [{ message: "\"anUndefinedVar\" is not defined.", type: "Identifier"}] },
         { code: "var a = b;", errors: [{ message: "\"b\" is not defined.", type: "Identifier"}] },
         { code: "function f() { b; }", errors: [{ message: "\"b\" is not defined.", type: "Identifier"}] },
-        { code: "/*global b:false*/ function f() { b = 1; }", errors: [{ message: "\"b\" is read only.", type: "Identifier"}] },
-        { code: "function f() { b = 1; }", global: { b: false }, errors: [{ message: "\"b\" is read only.", type: "Identifier"}] },
-        { code: "/*global b:false*/ function f() { b++; }", errors: [{ message: "\"b\" is read only.", type: "Identifier"}] },
-        { code: "/*global b*/ b = 1;", errors: [{ message: "\"b\" is read only.", type: "Identifier"}] },
-        { code: "/*global b:false*/ var b = 1;", errors: [{ message: "\"b\" is read only.", type: "Identifier"}] },
         { code: "window;", errors: [{ message: "\"window\" is not defined.", type: "Identifier"}] },
         { code: "require(\"a\");", errors: [{ message: "\"require\" is not defined.", type: "Identifier"}] },
-        { code: "Array = 1;", errors: [{ message: "\"Array\" is read only.", type: "Identifier"}] },
         { code: "var React; React.render(<img attr={a} />);", errors: [{ message: "\"a\" is not defined." }], ecmaFeatures: { jsx: true } },
         { code: "var React, App; React.render(<App attr={a} />);", errors: [{ message: "\"a\" is not defined." }], ecmaFeatures: { jsx: true } },
         { code: "[a] = [0];", ecmaFeatures: {destructuring: true}, errors: [{ message: "\"a\" is not defined." }] },

